src/components/graph/Graph.jsx

```diff
@@ -358,11 +358,74 @@ export default class Graph extends React.Component {
         this.pauseSimulation();
     }
 
+    _toggleNodeConnection(targetNodeId, allConnections) {
+        const newConnection = {
+            [targetNodeId]: allConnections[targetNodeId] === 1 ? 0 : 1
+        };
+
+        return { ...allConnections, ...newConnection };
+    }
+
+    _getLeafNodeConnections(startingNodeId, currentConnections) {
+        const startingNodeConnections = currentConnections[startingNodeId];
+        const startingNodeConnectionsList = Object.keys(startingNodeConnections);
+
+        return startingNodeConnectionsList.reduce((allLeafNodes, candidateLeafId) => {
+            const candidateLeafConnections = currentConnections[candidateLeafId];
+            const candidateLeafConnectionList = Object.keys(candidateLeafConnections);
+            const isLeafNode = candidateLeafConnectionList.length === 1;
+
+            if (isLeafNode) {
+                allLeafNodes[candidateLeafId] = candidateLeafConnections;
+            }
+
+            return allLeafNodes;
+        }, {});
+    }
+
+    onClickNode = clickedNodeId => {
+        const currentConnections = this.state.links;
```

Comment:
We could move all logic from line `387` to `411` in a separate function (maybe `collapseNodes`) the ouputs would be `toggledLeafNodes` and a new `d3Links` structure

Reply:
We could, and we should.

I also attempted this separation, so in the future this could be done to any node.
But it may need a few tweaks to be a bit more resilient for other nodes (not just leafs).

Reply:
I've moved this function for now into `graph.helper.js`. The discussion above could be a future concern :)